feat: Criação do cluster Nomad

nerdtux opened the pull request:
Utilizando os módulos IaC-Redes, IaC-Compute e imagem gerada pelo Packer.

Essa é uma versão inicial, existem issues geradas durante o teste
que precisam ser solucionadas.

Revisado por:
- Felipe Nobrega
- Jandson Oliveira
- Leo Barros
- Luiz Aoqui
- Marcos Rezende
- Marcelo Mansur
- Plinio
- Thais d'Angelis
- Victor Braga

Co-Author:
- Luiz Aoqui @lgfa29

Ref: #1

# Criação do cluster Nomad

<!-- Por favor descreva seu pull request aqui. -->

- [x] Garanta que seu **topic/feature/bugfix branch** tenha uma branch nomeada e não a sua branch main esteja no PR
- [x] Dê um titulo que expresse o objetivo do PR
- [x] Associe seu PR a uma Issue criada no repositósito. Caso seja uma correção de linguagem ou pequenas correções, não é necessário
- [x] Descreva o objetivo do PR
- [ ] Inclua links relevantes para a sua modificação/sugestão/correção
- [ ] Descreva um passo-a-passo para testar o seu PR

## Issue

#1

## Objetivo

Utilizar os módulos IaC-Redes, IaC-Compute e imagem gerada pelo Packer para criar o Cluster Nomad.

## Referências

## Como testar

<!-- Passo a passo -->

<!--
Marque um `x` dentro de [ ] para os itens que você forneceu informação
Para modificar este template no seu repositório, basta criar o arquivo .github/pull_request_template.md nele.
-->

nerdtux committed:
feat: Criação do cluster Nomad

Utilizando os módulos IaC-Redes, IaC-Compute e imagem gerada pelo Packer.

Essa é uma versão inicial, existem issues geradas durante o teste
que precisam ser solucionadas.

Revisado por:
- Felipe Nobrega
- Jandson Oliveira
- Leo Barros
- Luiz Aoqui
- Marcos Rezende
- Marcelo Mansur
- Plinio
- Thais d'Angelis
- Victor Braga

Co-Author:
- Luiz Aoqui @lgfa29

Ref: #1

nerdtux committed:
feat: Inclui for_each para adicionar vários clusters Nomad.

Utilizamos função for_each no arquivo main.tf buscando map
do arquivo terrraform.tfvars.

Signed-off-by: Gabriel Iglesias <ga12iglesias@gmail.com>
Co-authored-by: Danilo Figueiredo Rocha <snifbr@gmail.com>
Co-authored-by: Felipe Nobrega <lipenodias@gmail.com>
Co-authored-by: Marcelo Mansur <mansur.ufmg@gmail.com>

lgfa29 commented on main.tf:
```diff
@@ -2,25 +2,64 @@
+  for_each = var.server_pool
+
+  project                 = var.project
+  instance_name           = each.value.instance_name
```
```suggestion
  instance_name           = each.key
```
Aqui poderia usar a chave direto, já que ela também é usada para identificar an instância.

nerdtux replied:
Alterado!

lgfa29 commented on main.tf:
```diff
@@ -2,25 +2,64 @@
+  zone                    = each.value.zone
+  network                 = module.network_gcp.vpc_id
+  subnetwork              = module.network_gcp.subnets[0].id
+  tags                    = each.value.tag
+  metadata_startup_script = each.value.metadata_startup_script
+  labels                  = each.value.labels
```
Esses valores provavelmente vão ser sempre iguais para todas as instâncias, então não precisam ser definidos como variáveis.

nerdtux replied:
Optamos por colocar esses valores de forma dinâmica para futuramente termos a liberdade de, quando necessário, conseguirmos alterar apenas uma instância sem impactar todo o pool de recursos criados anteriormente.

gomex replied:
Seria legal termos labels padrões e a possibilidade de adicionar novas. O que acha?

lgfa29 commented on main.tf:
```diff
@@ -2,25 +2,64 @@
+}
+
+module "nomad_servers" {
+  source = "github.com/mentoriaiac/iac-modulo-compute-gcp.git?ref=v0.1.0"
```
```suggestion
  source = "github.com/mentoriaiac/iac-modulo-compute-gcp.git?ref=6ce607c"
```
Enquanto não sai o release novo, a gente pode usar esse [PR](https://github.com/mentoriaiac/iac-modulo-compute-gcp/pull/11) do @jandson-oliveira que adiciona a variável de service account.

Aí dentro dos módulos precisar colocar esse atributo:

```hcl
service_account_scopes = ["https://www.googleapis.com/auth/compute.readonly"]
```

nerdtux replied:
Alteração realizada!

gomex replied:
Release criada: https://github.com/mentoriaiac/iac-modulo-compute-gcp/releases/tag/v0.2.0

nerdtux replied:
Alterado!

lgfa29 commented on main.tf:
```diff
@@ -2,25 +2,64 @@
+module "nomad_clients" {
+  source = "github.com/mentoriaiac/iac-modulo-compute-gcp.git?ref=v0.1.0"
+
+  for_each = var.client_pool
```
Clients são máquinas sem estado, então acho que dá para tratar de forma mais efêmero, usando só o `count` mesmo. Assim todas as máquinas vão ser criadas iguais.

nerdtux replied:
Achamos que a sugestão faz sentido, mas gostaríamos de discutir em outra issue por ser uma decisão de conceito de implementação.

gomex replied:
Esse é um debate interessante. Vou abrir uma issue pra discutirmos isso lá: https://github.com/mentoriaiac/iac-nomad/issues/5

lgfa29 commented on variables.tf:
```diff
@@ -0,0 +1,36 @@
+  default     = "us-central1"
+}
+
+variable "server_pool" {
+  description = "Quantidade de Máquinas Servers"
+  type = map(object({
+    instance_name           = string
+    machine_type            = string
+    instance_image          = string
+    zone                    = string
+    metadata_startup_script = string
+    tag                     = list(string)
+    labels                  = map(string)
+  }))
+}
+
+variable "client_pool" {
+  description = "Quantidade de Máquinas Clients"
+  type = map(object({
+    instance_name           = string
+    machine_type            = string
+    instance_image          = string
+    zone                    = string
+    metadata_startup_script = string
+    tag                     = list(string)
+    labels                  = map(string)
+  }))
+}
```
Esse repositório guarda a configuração da infraestrutura, não um módulo. Então esse tipo de variável ficaria melhor como [`local`](https://www.terraform.io/docs/language/values/locals.html) já que o intuito não é controlar parâmetros de entrada, mas sim definir valores repetidos.

nerdtux replied:
Achamos que a sugestão faz sentido, mas gostaríamos de discutir em outra issue por ser uma decisão de conceito de implementação.

gomex replied:
@lgfa29 a sua ideia seria abandonar o uso de tfvars e usar apenas locals?

nerdtux replied:
Resolvido!

lgfa29 commented on main.tf:
```diff
@@ -2,25 +2,64 @@
+  network                 = module.network_gcp.vpc_id
+  subnetwork              = module.network_gcp.subnets[0].id
+  tags                    = each.value.tag
+  metadata_startup_script = each.value.metadata_startup_script
```
Usando o HEREDOC como alguém tinha mencionado na reunião funcionou 😄 

Para servidores:
```hcl
  metadata_startup_script = <<EOF
/usr/local/bin/nomad_bootstrap.sh server 3 '\"provider=gce project_name=${var.project} tag_value=nomad-server\"'
EOF
```

Para clientes:
```hcl
  metadata_startup_script = <<EOF
/usr/local/bin/nomad_bootstrap.sh client '\"provider=gce project_name=${var.project} tag_value=nomad-server\"'
EOF
```

nerdtux replied:
Alteração realizada!

lgfa29 commented on terraform.tfvars:
```diff
@@ -0,0 +1,58 @@
+    instance_image          = "orquestradores-1634167564"
+    zone                    = "us-central1-c"
+    metadata_startup_script = "echo 'nomad_boostrap.sh server 3' > /tmp/nomad_bootstrap.txt"
+    tag                     = ["nomad"]
```
As instâncias que vão rodar como servidor precisam ter a tag `nomad-server`.

nerdtux replied:
Alteração Realizada!

lgfa29 commented on .gitignore:
```diff
@@ -29,3 +29,6 @@ override.tf.json
 # example: *tfplan*
 tfplan
+key.json
+# terraform.tfvars
```
É bom continuar ignorando esse arquivo já que os valores vão ser diferentes para cada ambiente.

nerdtux replied:
Aguardando definição sobre a implementação da issue #4.

nerdtux replied:
Resolvido!

nerdtux committed:
refactor: Alterações realizadas à partir dos comentários recebidos no PR.

Correções nos arquivos de configuração.
Abrimos algumas issues para discussão de conceitos de implementação.
Iniciamos a configuração do arquivo de CI/CD, mas precisamos amadurecer
o conceito de utilizar o Makefile.

Signed-off-by: Gabriel Iglesias <ga12iglesias@gmail.com>
Co-Authored-by: Danilo Figueiredo Rocha <snifbr@gmail.com>
Co-authored-by: Felipe Nobrega <lipenodias@gmail.com>
Co-authored-by: Marcelo Mansur <mansur.ufmg@gmail.com>
Revisor: Diogo Lopes
Revisor: Gustavo Kuno
Revisor: Isaac Silva
Revisor: Plinio

gomex commented on main.tf:
```diff
@@ -2,25 +2,72 @@
+  region  = var.region
+}
+
+module "network_gcp" {
```
Por que estamos usando esse módulo aqui? Eu acredito que o groundwork poderia estar em outro pipeline. Se for algo apenas pra agilizar, ok ;)

snifbr replied:
Isso será resolvido como melhoria nesta Issue que foi aberta.
- https://github.com/mentoriaiac/iac-nomad/issues/7

gomex commented on main.tf:
```diff
@@ -2,25 +2,72 @@
+}
+
+module "nomad_servers" {
+  source = "github.com/mentoriaiac/iac-modulo-compute-gcp.git?ref=6ce607c"
```
Já pode usar a release v0.2.0: https://github.com/mentoriaiac/iac-modulo-compute-gcp/releases/tag/v0.2.0

nerdtux replied:
Resolvido!

gomex commented on main.tf:
```diff
@@ -2,25 +2,72 @@
+}
+
+module "nomad_clients" {
+  source = "github.com/mentoriaiac/iac-modulo-compute-gcp.git?ref=6ce607c"
```
Pode usar a release v0.2.0

nerdtux replied:
Resolvido!

gomex reviewed (changes requested):
Temos algumas coisas para avaliar, mas o PR está muito bom.

nerdtux committed:
refactor: Realizando alterações conforme comentários no PR.

Retiramos o .tfvars e alteramos para local.tf.
Atualizamos para a release v0.2.0 do módulo Compute GCP.
Alteramos o for_each dos clients para count.
Atualizamos e testamos a imagem gerada pelo módulo Packer.
Alteramos o nome dos servers para each.key (nome da instância = chave)

Signed-off-by: Gabriel Iglesias <ga12iglesias@gmail.com>
Co-Authored-by: Danilo Figueiredo Rocha <snifbr@gmail.com>
Co-Authored-by: Luiz Aoqui <lgfa29@gmail.com>
Co-authored-by: Felipe Nobrega <lipenodias@gmail.com>
Co-authored-by: Marcelo Mansur <mansur.ufmg@gmail.com>

gomex commented on local.tf:
```diff
@@ -0,0 +1,23 @@
+    server-node-1 = {
+      instance_name  = "server1"
+      machine_type   = "e2-small"
+      instance_image = "orquestradores-20211110235155"
```
Acho que por hora pode ser estático, mas no futuro isso precisa ser dinâmico via data ;)

snifbr replied:
Isso será resolvido como melhoria nesta Issue que foi aberta.
- https://github.com/mentoriaiac/iac-nomad/issues/6

gomex commented on local.tf:
```diff
@@ -0,0 +1,23 @@
+    server-node-2 = {
+      instance_name  = "server2"
+      machine_type   = "e2-medium"
+      instance_image = "orquestradores-20211110235155"
```
Acho que por hora pode ser estático, mas no futuro isso precisa ser dinâmico via data ;)

snifbr replied:
Isso será resolvido como melhoria nesta Issue que foi aberta.
- https://github.com/mentoriaiac/iac-nomad/issues/6

gomex commented on local.tf:
```diff
@@ -0,0 +1,23 @@
+    server-node-3 = {
+      instance_name  = "server3"
+      machine_type   = "e2-medium"
+      instance_image = "orquestradores-20211110235155"
```
Acho que por hora pode ser estático, mas no futuro isso precisa ser dinâmico via data ;)

snifbr replied:
Isto será resolvido como melhoria nesta Issue que foi aberta.
- https://github.com/mentoriaiac/iac-nomad/issues/6

gomex commented on main.tf:
```diff
@@ -2,25 +2,89 @@
+  project                 = var.project
+  instance_name           = "client-${count.index + 1}"
+  machine_type            = "e2-medium"
+  instance_image          = "orquestradores-20211110235155"
```
Acho que por hora pode ser estático, mas no futuro isso precisa ser dinâmico via data ;)

snifbr replied:
Isso será resolvido como melhoria nesta Issue que foi aberta.
- https://github.com/mentoriaiac/iac-nomad/issues/6

gomex commented on variables.tf:
```diff
@@ -0,0 +1,11 @@
+variable "project" {
+  description = "Nome do projeto (Default é staging)"
```
Se o default é staging, pq não tem default aqui?

snifbr replied:
Isso será resolvido logo em seguida como um simples fix no próximo fork para lançarmos a primeira release.

gomex commented on .github/workflows/ci.yml:
```diff
@@ -36,17 +35,70 @@ jobs:
+        run: terraform plan -out="terraform_plan" -no-color
+        if: github.event_name == 'pull_request'
+
+      - name: Plan output
```
Vou ignorar o ci/cd por hora, pois vamos discutir ele melhor depois, né?

snifbr replied:
OK, iremos discutir posteriormente.

gomex reviewed (commented):
Está muito bom o PR, mas pontuei algumas coisas, que podem ser ignoradas, caso não faça sentido.

snifbr reviewed (approved):
LGTM
